### ISSUE TITLE
Keyboard input methods that allow for multiple characters only expose first character to event.getKey()

### ISSUE THREAD

**Issue by smathot** [psychopy/psychopy#3318]
Certain input methods allow for multiple characters to be composed at once, and then these are sent at once as an instantaneous sequence of key presses to the program. For example, Chinese Pinyin input allows you to enter a string of latin characters (nihao), which are then mapped onto the corresponding Chinese characters (你好) which are then sent to the software. I believe that many complex input methods, for example for Japanese, use the same general principle.

The desired behavior of `event.getKey()` would be that either all characters are returned as a list of individual characters (`['你', '好']`) or as a list that is grouped by input (`['你好']`). Instead, `event.getKey()` only returns the first character (`['你']`) and the rest appears to be discarded.

I experienced a similar issue with PyGame, and in that case the issue was due to inappropriate clearing of the event queue. See also:

- open-cogsci/opensesame#736

**Comment by TEParsons** [comment]
That's the intended behaviour of `getKey`, to get all keys you can call `getKeys` :)

**Comment by smathot** [comment]
I actually meant `getKeys()` :-) (As far as I know, there is no `getKey()` function?)

**Comment by TEParsons** [comment]
Ah I see, I just tested it now and what you describe is exactly how it does work - it shows every key pressed since the last call to `event.getKeys()` as a list. Could you send the code you were running to get this error?

**Comment by peircej** [comment]
> I actually meant getKeys() :-) (As far as I know, there is no getKey() function?)

hahaha, it also had me confused - I had to go and check if we had a method I wasn't aware of! ;-)

OK, TL;DR: `event.getKeys()` seems fine, as does `Keyboard` class when using the psychtoolbox engine, but the event fallback for the Keyboard class is doing as @smathot says - it returns just the first key pressed.

Using event.getKeys() works correctly:
```
from psychopy import visual, event
win = visual.Window()
for frame in range(120):
    win.flip() # press several keys
print(f'keys: {event.getKeys()}')
```

Then I tested the new hardware.keyboard.Keyboard class:
```
from psychopy import visual
from psychopy.hardware import keyboard

win = visual.Window()
kb = keyboard.Keyboard()
for frame in range(120):
    win.flip() # press several keys

for key in kb.getKeys():
    print(key.name, key.code)
```
and that gave me a single character!

Then I realised this laptop didn't have the psychtoolbox extension active (`WARNING     Import Error: No module named 'psychtoolbox'. Using event module for keyboard component.`) so this was using Keyboard class but _falling back_ to `event` within that class.

I installed psychtoolbox and got a list of all keys pressed again.

**Comment by smathot** [comment]
The first code, using `event.getKeys()` works with Latin input:

```
keys: ['t', 'e', 's', 't', 'i', 'n', 'g']
```

But for Pinyin input it only keeps the first character. Below is what I get when I commit the string 你好 at once:

```
keys: ['你']
```

The second code, using `kb.getKeys()`, also works with Latin input. But with Pinyin input it just returns the Latin characters that are supposed to be translated to Chinese. ni = 你, hao = 好. The 1 at the end is actually the character that should commit the string. (You get a popup box with several character strings that 'nihao' could correspond to, because a lot of characters are spelled the same in Pinyin. And I selected alternative 1.)

```
n 78
i 73
h 72
a 65
o 79
1 49
```

Long story short: This is not only about being able to accept multiple characters, although that is a prerequisite. It's also about dealing with exotic input methods!

**Comment by m-macaskill** [comment]
Doesn't this require require responding to `pyglet`'s `on_text` event rather than its key events? I think that explicitly returns multi-character text, not just for complex scripts like this, but also other alternative text input methods (e.g. handwriting recognition, which can return whole words simultaneously, with no key being pushed at all)?

**Comment by peircej** [comment]
Yes, exactly as Mike says. `event.getKeys()` is not designed to _interpret_ the characters of the keys pressed - it merely reports the names of the keys that were pressed. If you press shift and a you don't get A, you get ['shift', 'a']. That's the way to get fastest response times. 

To get characters, instead of key names, you should capture events from pyglet.on_text. The way PsychoPy makes use of that is to pass all events to the PygletBackend.onText and then looks for the currently focused editable (TextBox2) object and passes the characters to that. These should work together, so you can simultaneously get the key names and low-latency RTs from the harware keyboard, as well as representing the text being typed in the editable TextBox2 object.

@smathot if you want access to the chars you could do one of:
- monkey-patch the Window.backend.onText function with your own function
- redirect pyglet.on_text

If you're ultimately trying to create an editable text box then maybe try out out TextBox2. We recreated text rendering from scratch because pyglet and pygame are both slow to update text for rendering. TextBox2 can layout and render thousands of chars on every screen refresh without dropping a frame. I can chat (email) about how that differs from the other methods and what the caveats are

**Comment by smathot** [comment]
@peircej Thanks for the suggesting to monkey-patch the `onText()`. (That might solve my specific case.) I see the argument that the `getKeys()` function is basically for capturing key-press responses as efficiently as possible, and not for implementing a word processor. Would it make sense to indicate this clearly in the documentation, and then perhaps offer an additional `getText()` function?
